apps/emqx_durable_storage/src/emqx_ds_message_storage_bitmask.erl

```diff
@@ -0,0 +1,731 @@
+    read_options = [] :: emqx_ds_storage_layer:db_write_options()
+}).
+
+-record(it, {
+    handle :: rocksdb:itr_handle(),
+    filter :: keyspace_filter(),
+    cursor :: binary() | undefined,
+    next_action :: {seek, binary()} | next,
+    refresh_counter :: {non_neg_integer(), pos_integer()} | undefined
+}).
+
+-record(filter, {
+    keymapper :: keymapper(),
+    topic_filter :: emqx_topic:words(),
+    start_time :: integer(),
+    hash_bitfilter :: integer(),
+    hash_bitmask :: integer(),
+    time_bitfilter :: integer(),
+    time_bitmask :: integer()
+}).
+
+% NOTE
+% Keymapper decides how to map messages into RocksDB column family keyspace.
+-record(keymapper, {
+    source :: [bitsource(), ...],
+    bitsize :: bits(),
+    epoch :: non_neg_integer()
+}).
+
+-type bitsource() ::
+    %% Consume `_Size` bits from timestamp starting at `_Offset`th bit.
+    %% TODO consistency
+    {timestamp, _Offset :: bits(), _Size :: bits()}
+    %% Consume next topic level (either one or all of them) and compute `_Size` bits-wide hash.
+    | {hash, level | levels, _Size :: bits()}.
+
+-opaque db() :: #db{}.
+-opaque iterator() :: #it{}.
+-type keymapper() :: #keymapper{}.
+-type keyspace_filter() :: #filter{}.
+
+%%================================================================================
+%% API funcions
+%%================================================================================
+
+%% Create a new column family for the generation and a serializable representation of the schema
+-spec create_new(rocksdb:db_handle(), emqx_ds_storage_layer:gen_id(), options()) ->
+    {schema(), emqx_ds_storage_layer:cf_refs()}.
+create_new(DBHandle, GenId, Options) ->
+    CFName = data_cf(GenId),
+    CFOptions = maps:get(cf_options, Options, []),
+    {ok, CFHandle} = rocksdb:create_column_family(DBHandle, CFName, CFOptions),
+    Schema = #schema{keymapper = make_keymapper(Options)},
+    {Schema, [{CFName, CFHandle}]}.
+
+%% Reopen the database
+-spec open(
+    emqx_ds:shard(),
+    rocksdb:db_handle(),
+    emqx_ds_storage_layer:gen_id(),
+    emqx_ds_storage_layer:cf_refs(),
+    schema()
+) ->
+    db().
+open(Shard, DBHandle, GenId, CFs, #schema{keymapper = Keymapper}) ->
+    {value, {_, CFHandle}} = lists:keysearch(data_cf(GenId), 1, CFs),
+    #db{
+        shard = Shard,
+        handle = DBHandle,
+        cf = CFHandle,
+        keymapper = Keymapper
+    }.
+
+-spec make_keymapper(options()) -> keymapper().
+make_keymapper(#{
+    timestamp_bits := TimestampBits,
+    topic_bits_per_level := BitsPerLevel,
+    epoch := MaxEpoch
+}) ->
+    TimestampLSBs = min(TimestampBits, floor(math:log2(MaxEpoch))),
+    TimestampMSBs = TimestampBits - TimestampLSBs,
+    NLevels = length(BitsPerLevel),
+    {LevelBits, [TailLevelsBits]} = lists:split(NLevels - 1, BitsPerLevel),
+    Source = lists:flatten([
+        [{timestamp, TimestampLSBs, TimestampMSBs} || TimestampMSBs > 0],
```

Comment:
Possibly a dumb question: if `TimestampLSBs` is an [offset](https://github.com/emqx/emqx/pull/11114#event-9605545761), shouldn't this be the other way around?  i.e.: `{timestamp, MSBs, LSBs}` (skip the MSBs and take the LSBs)?

Reply:
Heh, I could probably understand the confusion. So, here we're saying "take `TimestampMSBs` bits starting for `TimestampLSBs`th one". But that bitsource actually denotes part of a bitstring laying in the MSBs region, i.e. for consistency we'd need to reverse this list. It's was just easier to deal with this list in reversed order IIRC.